hazelcast/src/main/java/com/hazelcast/internal/networking/ChannelHandler.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.networking;
+
+
+/**
+ * ChannelHandler is either responsible for processing inbound or outbound data.
+ *
+ * ChannelHandlers are not expected to be thread-safe and the same instance can't
+ * shared over multiple pipelines.
+ *
+ *
+ * <h1>Nio read spinloop</h1>
+ *
+ *
+ * <h1>io read deadlock</h1>
+ *
+ * What happens when data is read from the socket into some buffer, but in a
+ * subsequent buffer there is no space to write everything. If there is no data
+ * in the socket, then the connection will 'deadlock' because it will not receive
+ * any read events.
+ *
+ * <h1>Replacing handlers and pending data</h1>
+ *
+ * If an outbound handler gets replaced, but has pending data, this data is
+ * problematic. An example of this is protocol and tls handshake, when
+ * ProtocolEncoder replaced itself by a PacketEncoder, it still has the actual
+ * protocol to write. But this should not be dealt with by the TldHandshakeEncoder,
+ * because it will mess up the communication with the other side.
+ *
+ * Perhaps currently not a problem because in case of the ProtocolEncoder, it
+ * will copy the data to the PacketHandler, not to the TlsHandshakeEncoder.
+ *
+ * Replacing the source/destination of a Handler after the Handler is registered
+ * is not adviced, it can lead to src/destination changes not being picked up
+ * because the linking process will not be repeated.
+ *
+ *
+ *
+ *
+ * @param <S> the type of the source. E.g. a ByteBuffer or
+ *           a {@link com.hazelcast.util.function.Supplier}.
+ * @param <D> the type of the destination. E.g. a ByteBuffer
+ *           or a {@link com.hazelcast.util.function.Consumer}.
+ */
+public abstract class ChannelHandler<H extends ChannelHandler, S, D> {
+
+    /**
+     * The Channel this InboundHandler is connected to. Only the {
+     * @link ChannelInboundPipeline} should modify this field.
+     */
+    protected Channel channel;
```

Comment:
Is this `channel` field only for inbound pipeline? not valid for outbound? If yes, shouldn't this moved to `ChannelInboundHandler `?

Reply:
Fixed